### PR TITLE
add multi-arch (arm64) support for docker/podman drivers

### PR DESCRIPTION
**Pull request by ilya-zuyev**
This PR changes kicbase image to multi-arch format, supporting arm64/linux architecture and enables docker driver for arm64 systems

In the example below I'm using `ilyazuyev/kicbase:pr-test` image name for demo purposes

```
ilyaz@skeletron --- » KICBASE_IMAGE_REGISTRIES=ilyazuyev/kicbase:pr-test make push-kic-base-image

env DOCKER_CLI_EXPERIMENTAL=enabled docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
Setting /usr/bin/qemu-alpha-static as binfmt interpreter for alpha
Setting /usr/bin/qemu-arm-static as binfmt interpreter for arm
Setting /usr/bin/qemu-armeb-static as binfmt interpreter for armeb
Setting /usr/bin/qemu-sparc-static as binfmt interpreter for sparc
Setting /usr/bin/qemu-sparc32plus-static as binfmt interpreter for sparc32plus
Setting /usr/bin/qemu-sparc64-static as binfmt interpreter for sparc64
Setting /usr/bin/qemu-ppc-static as binfmt interpreter for ppc
Setting /usr/bin/qemu-ppc64-static as binfmt interpreter for ppc64
Setting /usr/bin/qemu-ppc64le-static as binfmt interpreter for ppc64le
Setting /usr/bin/qemu-m68k-static as binfmt interpreter for m68k
Setting /usr/bin/qemu-mips-static as binfmt interpreter for mips
Setting /usr/bin/qemu-mipsel-static as binfmt interpreter for mipsel
Setting /usr/bin/qemu-mipsn32-static as binfmt interpreter for mipsn32
Setting /usr/bin/qemu-mipsn32el-static as binfmt interpreter for mipsn32el
Setting /usr/bin/qemu-mips64-static as binfmt interpreter for mips64
Setting /usr/bin/qemu-mips64el-static as binfmt interpreter for mips64el
Setting /usr/bin/qemu-sh4-static as binfmt interpreter for sh4
Setting /usr/bin/qemu-sh4eb-static as binfmt interpreter for sh4eb
Setting /usr/bin/qemu-s390x-static as binfmt interpreter for s390x
Setting /usr/bin/qemu-aarch64-static as binfmt interpreter for aarch64
Setting /usr/bin/qemu-aarch64_be-static as binfmt interpreter for aarch64_be
Setting /usr/bin/qemu-hppa-static as binfmt interpreter for hppa
Setting /usr/bin/qemu-riscv32-static as binfmt interpreter for riscv32
Setting /usr/bin/qemu-riscv64-static as binfmt interpreter for riscv64
Setting /usr/bin/qemu-xtensa-static as binfmt interpreter for xtensa
Setting /usr/bin/qemu-xtensaeb-static as binfmt interpreter for xtensaeb
Setting /usr/bin/qemu-microblaze-static as binfmt interpreter for microblaze
Setting /usr/bin/qemu-microblazeel-static as binfmt interpreter for microblazeel
Setting /usr/bin/qemu-or1k-static as binfmt interpreter for or1k
env DOCKER_CLI_EXPERIMENTAL=enabled docker buildx rm --builder minikube-builder || true
error: no builder "minikube-builder" found


env DOCKER_CLI_EXPERIMENTAL=enabled docker buildx create --name kicbase-builder --buildkitd-flags '--debug' --use || true
error: existing instance for kicbase-builder but no append mode, specify --node to make changes for existing instances
Error response from daemon: manifest for ilyazuyev/kicbase:pr-test not found: manifest unknown: manifest unknown
Image doesn't exist in registry

⚠️  'Are you sure you want to push ilyazuyev/kicbase:pr-test ?'
Do you want to proceed? (Y/N): y
env DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --platform linux/arm64,linux/amd64 -t ilyazuyev/kicbase:pr-test --push  --build-arg COMMIT_SHA=v1.16.0-"9c4a76581c9c2562feb204a108eeb09f5c54e6a9" ./deploy/kicbase
[+] Building 10.0s (47/70)                                                                          
 => [internal] load build definition from Dockerfile                                           0.0s
 => => transferring dockerfile: 32B                                                            0.0s
 => [internal] load .dockerignore                                                              0.0s
 => => transferring context: 2B                                                                0.0s
 => [linux/amd64 internal] load metadata for docker.io/library/ubuntu:focal-20201106           0.9s
 => [auth] library/ubuntu:pull token for registry-1.docker.io                                  0.0s
 => [linux/arm64 internal] load metadata for docker.io/library/ubuntu:focal-20201106           1.7s
 => [internal] load build context                                                              0.0s
 => => transferring context: 443B                                                              0.0s
 => [linux/arm64  1/32] FROM docker.io/library/ubuntu:focal-20201106@sha256:c95a8e48bf88e9849  0.0s
 => => resolve docker.io/library/ubuntu:focal-20201106@sha256:c95a8e48bf88e9849f3e0f723d9f49f  0.0s
 => [linux/amd64  1/32] FROM docker.io/library/ubuntu:focal-20201106@sha256:c95a8e48bf88e9849  0.0s
 => => resolve docker.io/library/ubuntu:focal-20201106@sha256:c95a8e48bf88e9849f3e0f723d9f49f  0.0s
 => CACHED [linux/arm64  2/32] COPY 10-network-security.conf /etc/sysctl.d/10-network-securit  0.0s
 => CACHED [linux/arm64  3/32] COPY 11-tcp-mtu-probing.conf /etc/sysctl.d/11-tcp-mtu-probing.  0.0s
 => CACHED [linux/arm64  4/32] COPY clean-install /usr/local/bin/clean-install                 0.0s
 => CACHED [linux/arm64  5/32] COPY entrypoint /usr/local/bin/entrypoint                       0.0s
 => CACHED [linux/arm64  6/32] RUN echo "Ensuring scripts are executable ..."     && chmod +x  0.0s
 => CACHED [linux/arm64  7/32] RUN clean-install     lz4     gnupg     sudo     openssh-serve  0.0s
 => CACHED [linux/arm64  8/32] RUN sh -c "echo 'deb https://download.docker.com/linux/ubuntu   0.0s
 => CACHED [linux/arm64  9/32] RUN export ARCH=$(dpkg --print-architecture | sed 's/ppc64el/p  0.0s
 => [linux/arm64 10/32] RUN sh -c "echo 'deb https://download.opensuse.org/repositories/devel  8.1s
 => => # Get:3 https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbun
 => => # tu_20.04  InRelease [1634 B]                                                              
 => => # Get:4 http://ports.ubuntu.com/ubuntu-ports focal-updates InRelease [114 kB]              
 => => # Get:5 http://ports.ubuntu.com/ubuntu-ports focal-backports InRelease [101 kB]            
 => => # Get:6 http://ports.ubuntu.com/ubuntu-ports focal-security InRelease [109 kB]              
 => => # Get:7 https://download.docker.com/linux/ubuntu focal/stable arm64 Packages [6174 B]      
 => CACHED [linux/amd64  2/32] COPY 10-network-security.conf /etc/sysctl.d/10-network-securit  0.0s
 => CACHED [linux/amd64  3/32] COPY 11-tcp-mtu-probing.conf /etc/sysctl.d/11-tcp-mtu-probing.  0.0s
 => CACHED [linux/amd64  4/32] COPY clean-install /usr/local/bin/clean-install                 0.0s
 => CACHED [linux/amd64  5/32] COPY entrypoint /usr/local/bin/entrypoint                       0.0s
 => CACHED [linux/amd64  6/32] RUN echo "Ensuring scripts are executable ..."     && chmod +x  0.0s
 => CACHED [linux/amd64  7/32] RUN clean-install     lz4     gnupg     sudo     openssh-serve  0.0s
 => CACHED [linux/amd64  8/32] RUN sh -c "echo 'deb https://download.docker.com/linux/ubuntu   0.0s
 => CACHED [linux/amd64  9/32] RUN export ARCH=$(dpkg --print-architecture | sed 's/ppc64el/p  0.0s
 => CACHED [linux/amd64 10/32] RUN sh -c "echo 'deb https://download.opensuse.org/repositorie  0.0s
 => CACHED [linux/amd64 11/32] RUN sh -c "echo 'deb https://download.opensuse.org/repositorie  0.0s
 => CACHED [linux/amd64 12/32] RUN sh -c "echo 'deb http://download.opensuse.org/repositories  0.0s
 => CACHED [linux/amd64 13/32] COPY automount/minikube-automount /usr/sbin/minikube-automount  0.0s
 => CACHED [linux/amd64 14/32] COPY automount/minikube-automount.service /usr/lib/systemd/sys  0.0s
 => CACHED [linux/amd64 15/32] RUN ln -fs /usr/lib/systemd/system/minikube-automount.service   0.0s
 => CACHED [linux/amd64 16/32] COPY scheduled-stop/minikube-scheduled-stop /var/lib/minikube/  0.0s
 => CACHED [linux/amd64 17/32] COPY scheduled-stop/minikube-scheduled-stop.service /usr/lib/s  0.0s
 => CACHED [linux/amd64 18/32] RUN ln -fs /usr/lib/systemd/system/minikube-scheduled-stop.ser  0.0s
 => CACHED [linux/amd64 19/32] RUN systemctl disable containerd && systemctl disable crio &&   0.0s
 => CACHED [linux/amd64 20/32] RUN systemctl enable docker.service && systemctl enable podman  0.0s
 => CACHED [linux/amd64 21/32] RUN mkdir /var/run/sshd                                         0.0s
 => CACHED [linux/amd64 22/32] RUN echo 'root:root' |chpasswd                                  0.0s
 => CACHED [linux/amd64 23/32] RUN sed -ri 's/^#?PermitRootLogin\s+.*/PermitRootLogin yes/' /  0.0s
 => CACHED [linux/amd64 24/32] RUN sed -ri 's/UsePAM yes/#UsePAM yes/g' /etc/ssh/sshd_config   0.0s
 => CACHED [linux/amd64 25/32] RUN sed -ri 's/dns files/files dns/g' /etc/nsswitch.conf        0.0s
 => CACHED [linux/amd64 26/32] RUN adduser --ingroup docker --disabled-password --gecos '' do  0.0s
 => CACHED [linux/amd64 27/32] RUN adduser docker sudo                                         0.0s
 => CACHED [linux/amd64 28/32] RUN adduser docker podman                                       0.0s
 => CACHED [linux/amd64 29/32] RUN echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers         0.0s
 => CACHED [linux/amd64 30/32] RUN mkdir /home/docker/.ssh                                     0.0s
 => CACHED [linux/amd64 31/32] RUN mkdir -p /kind                                              0.0s
 => CACHED [linux/amd64 32/32] RUN rm -rf   /usr/share/doc/*   /usr/share/man/*   /usr/share/  0.0s
 
 [skip]
 
  => exporting to image                                                                       789.0s
 => => exporting layers                                                                       20.1s
 => => exporting manifest sha256:6259cc1fcbddc89344e698c2143a3319891fb0f86eec7ff39c0cc2fa1cc3  0.0s
 => => exporting config sha256:a2513a598ee67b204ba18107590dd3226f0e83c26514ab708a6bd88cd88de6  0.0s
 => => exporting manifest sha256:d294dec113c9da7948e063b0bc609ba97b35de48aa47d9eff9d9686fbd12  0.0s
 => => exporting config sha256:989c8c2a18b426c83c33be861bff77b131424d02397ac870db18d0ee0ad4ca  0.0s
 => => exporting manifest list sha256:b31df643a3636a09b67cacd8dc8fbf4700e6ff3831bc7f33eee15f3  0.0s
 => => pushing layers                                                                        762.5s
 => => pushing manifest for docker.io/ilyazuyev/kicbase:pr-test                             6.1s
 => [auth] ilyazuyev/kicbase:pull,push token for registry-1.docker.io                          0.0s
 => [auth] ilyazuyev/kicbase:pull,push token for registry-1.docker.io                          0.0s
```

Then on an arm64 machine:
```
ubuntu@ip-172-31-39-30:~$ uname -a
Linux ip-172-31-39-30 5.4.0-1032-aws #33-Ubuntu SMP Wed Dec 9 17:17:06 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
```

```
ubuntu@ip-172-31-39-30:~/src/minikube$ time ./out/minikube start --driver=docker --base-image=ilyazuyev/kicbase:pr-test
* minikube v1.16.0 on Ubuntu 20.04 (arm64)
* Using the docker driver based on user configuration
* Starting control plane node minikube in cluster minikube
* Creating docker container (CPUs=2, Memory=2200MB) .../ E0108 21:05:57.103205  411350 cache.go:63] save image to file "k8s.gcr.io/etcd-arm64:3.4.13-0" -> "/home/ubuntu/.minikube/cache/images/k8s.gcr.io/etcd-arm64_3.4.13-0" failed: nil image for k8s.gcr.io/etcd-arm64:3.4.13-0: GET https://k8s.gcr.io/v2/etcd-arm64/manifests/3.4.13-0: MANIFEST_UNKNOWN: Failed to fetch "3.4.13-0" from request "/v2/etcd-arm64/manifests/3.4.13-0".
- E0108 21:06:05.577136  411350 cache.go:194] Error caching images:  Caching images for kubeadm: caching images: caching image "/home/ubuntu/.minikube/cache/images/k8s.gcr.io/etcd-arm64_3.4.13-0": nil image for k8s.gcr.io/etcd-arm64:3.4.13-0: GET https://k8s.gcr.io/v2/etcd-arm64/manifests/3.4.13-0: MANIFEST_UNKNOWN: Failed to fetch "3.4.13-0" from request "/v2/etcd-arm64/manifests/3.4.13-0".

* Preparing Kubernetes v1.20.0 on Docker 20.10.2 ...| X Unable to load cached images: loading cached images: stat /home/ubuntu/.minikube/cache/images/k8s.gcr.io/etcd-arm64_3.4.13-0: no such file or directory
    > kubectl.sha256: 64 B / 64 B [--------------------------] 100.00% ? p/s 0s
    > kubelet.sha256: 64 B / 64 B [--------------------------] 100.00% ? p/s 0s
    > kubeadm.sha256: 64 B / 64 B [--------------------------] 100.00% ? p/s 0s
    > kubeadm: 34.50 MiB / 34.50 MiB [----------------] 100.00% 6.07 MiB p/s 6s
    > kubectl: 35.44 MiB / 35.44 MiB [---------------] 100.00% 3.05 MiB p/s 12s
    > kubelet: 100.53 MiB / 100.53 MiB [-------------] 100.00% 2.97 MiB p/s 34s

  - Generating certificates and keys ...
  - Booting up control plane ...
  - Configuring RBAC rules ...
* Verifying Kubernetes components...
* Enabled addons: storage-provisioner, default-storageclass
* Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

real    1m33.658s
user    0m9.940s
sys     0m3.225s
ubuntu@ip-172-31-39-30:~/src/minikube$ 
```

```
ubuntu@ip-172-31-39-30:~/src/minikube$ docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED              STATUS              PORTS                                                                                                      NAMES
35651baf3934        ilyazuyev/kicbase:pr-test   "/usr/local/bin/entr…"   About a minute ago   Up About a minute   127.0.0.1:32811->22/tcp, 127.0.0.1:32810->2376/tcp, 127.0.0.1:32809->5000/tcp, 127.0.0.1:32808->8443/tcp   minikube
ubuntu@ip-172-31-39-30:~/src/minikube$ 
```

```
ubuntu@ip-172-31-39-30:~/src/minikube$ kubectl get pods --all-namespaces
NAMESPACE     NAME                               READY   STATUS    RESTARTS   AGE
kube-system   coredns-74ff55c5b-tbdb2            1/1     Running   0          58s
kube-system   etcd-minikube                      1/1     Running   0          70s
kube-system   kube-apiserver-minikube            1/1     Running   0          70s
kube-system   kube-controller-manager-minikube   1/1     Running   0          70s
kube-system   kube-proxy-jrpht                   1/1     Running   0          58s
kube-system   kube-scheduler-minikube            1/1     Running   0          70s
kube-system   storage-provisioner                1/1     Running   0          73s
ubuntu@ip-172-31-39-30:~/src/minikube$ 
```

### CODE REVIEW

**Comment by tstromberg** [comment]
Any reason for the name change?

**Comment by ilya-zuyev** [comment]
> Any reason for the name change?

The idea is to support both multi-arch kicbase image and the amd64 only for some time.
Docker buildkit support is still kind of experimental feature, just to avoid potential problems with old docker version on client machines we try to use `kicbase-multiarch` and if it can't be used we fallback to the amd64 `kicbase`

**Comment by tstromberg** [comment]
AFAIK, buildx is experimental, but multi-arch pulls should not be experimental any longer. Docker made it's images multi-arch in 2017: https://www.docker.com/blog/docker-official-images-now-multi-platform/

Kubernetes dropped suffixes for architectures about 2 years ago. I would suggest we do the same until a specific limitation can be identified.

**Comment by afbjorklund** [comment]
What we (and they) did for the other images, is that we used the arch suffix for the specific images.

So you would have a "foo-amd64" and an "foo-arm64", and then a manifest for those called "foo".

**Comment by afbjorklund** [comment]
> AFAIK, buildx is experimental, but multi-arch pulls should not be experimental any longer.

The main problem is that the _daemon_ doesn't understand architectures.

So you can only work with one architecture at the time, when cross-building.
If you need to do something else, you need to work with a (local?) registry.

I did that workaround in my multi-image build, when kic inherited from kind...

**Comment by ilya-zuyev** [comment]
> AFAIK, buildx is experimental, but multi-arch pulls should not be experimental any longer. Docker updated all of it's images to multi-arch back in 2017: https://www.docker.com/blog/docker-official-images-now-multi-platform/
> 
> Kubernetes dropped suffixes for architectures about 2 years ago. To make the code simpler, I would suggest we do the same until a specific limitation can be identified.

ok, makes sense. will fix.

File: Makefile [hunk 1]
```diff
@@ -600,6 +600,27 @@ kic-base-image: ## builds the kic base image and tags local/kicbase:latest and l
 	docker tag local/kicbase:$(KIC_VERSION) local/kicbase:latest
 	docker tag local/kicbase:$(KIC_VERSION) local/kicbase:$(KIC_VERSION)-$(COMMIT_SHORT)
 
+# multi-arch docker images
+X_DOCKER_BUILDER ?= minikube-builder
+X_BUILD_ENV ?= DOCKER_CLI_EXPERIMENTAL=enabled
+
+.PHONY: docker-multi-arch-builder
+docker-multi-arch-builder:
```
Comment (medyagh) [inline Makefile:608]: lets ensure it doesn't over-write the image

File: Makefile [hunk 2]
```diff
@@ -600,6 +600,27 @@ kic-base-image: ## builds the kic base image and tags local/kicbase:latest and l
...
+
+.PHONY: kic-base-image-multi-arch
+kic-base-image-multi-arch: docker-multi-arch-builder
+ifdef AUTOPUSH
+	docker login gcr.io/k8s-minikube
+	docker login docker.pkg.github.com
+	docker login
+endif
+	env $(X_BUILD_ENV) docker buildx build --platform $(KICBASE_ARCH) $(addprefix -t ,$(MA_KICBASE_IMAGE_REGISTRY)) --push  --build-arg COMMIT_SHA=${VERSION}-$(COMMIT) ./deploy/kicbase
+
```
Comment (medyagh) [inline Makefile:623]: lets comment out the previous push image implemntations

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/drivers/kic/types.go [hunk 3]
```diff
@@ -26,15 +26,29 @@ const (
 	// Version is the current version of kic
 	Version = "v0.0.15-snapshot4"
 	// SHA of the kic base image
-	baseImageSHA = "ef1f485b5a1cfa4c989bc05e153f0a8525968ec999e242efff871cbb31649c16"
+	baseImageSHA          = "ef1f485b5a1cfa4c989bc05e153f0a8525968ec999e242efff871cbb31649c16"
+	baseMultiArchImageSHA = "todo"
```
Comment (medyagh) [inline pkg/drivers/kic/types.go:30]: get rid of this line

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/drivers/kic/types.go [hunk 4]
```diff
@@ -26,15 +26,29 @@ const (
 	// Version is the current version of kic
 	Version = "v0.0.15-snapshot4"
 	// SHA of the kic base image
-	baseImageSHA = "ef1f485b5a1cfa4c989bc05e153f0a8525968ec999e242efff871cbb31649c16"
+	baseImageSHA          = "ef1f485b5a1cfa4c989bc05e153f0a8525968ec999e242efff871cbb31649c16"
```
Comment (medyagh) [inline pkg/drivers/kic/types.go:29]: lets verify if SHA is same for arm64 and amd if not, we should have two shas in the code

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/drivers/kic/types.go (hunk 4 above)
Comment (ilya-zuyev) [reply pkg/drivers/kic/types.go]:
```
ubuntu@ip-172-31-39-30:~/src/minikube$ uname -a 
Linux ip-172-31-39-30 5.4.0-1032-aws #33-Ubuntu SMP Wed Dec 9 17:17:06 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@ip-172-31-39-30:~/src/minikube$ docker inspect --format='{{index .RepoDigests 0}}' ilyazuyev/kicbase:pr-test
ilyazuyev/kicbase@sha256:949338df4b157b907fbc9b27466c160c609f2f0f17c5dce73d2ceda084fedb69
```

```
izuyev@izuyev-macbookpro --- ~ » uname -a
Darwin izuyev-macbookpro.roam.corp.google.com 19.6.0 Darwin Kernel Version 19.6.0: Thu Oct 29 22:56:45 PDT 2020; root:xnu-6153.141.2.2~1/RELEASE_X86_64 x86_64
izuyev@izuyev-macbookpro --- ~ » docker inspect --format='{{index .RepoDigests 0}}' ilyazuyev/kicbase:pr-test
ilyazuyev/kicbase@sha256:949338df4b157b907fbc9b27466c160c609f2f0f17c5dce73d2ceda084fedb69
izuyev@izuyev-macbookpro --- ~ » 
```

It's the same sha1

File: pkg/drivers/kic/types.go (hunk 3 above)
Comment (ilya-zuyev) [reply pkg/drivers/kic/types.go]: done.

File: pkg/drivers/kic/types.go [hunk 5]
```diff
@@ -24,7 +24,7 @@ import (
 
 const (
 	// Version is the current version of kic
-	Version = "v0.0.15-snapshot4"
+	Version = "v0.0.16-snapshot1"
```
Comment (medyagh) [inline pkg/drivers/kic/types.go:27]: lets update the SHA as well

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/drivers/kic/types.go (hunk 5 above)
Comment (ilya-zuyev) [reply pkg/drivers/kic/types.go:27]: done

File: Makefile [hunk 6]
```diff
@@ -648,8 +675,8 @@ push-kic-base-image-hub: kic-base-image ## Push kic-base to docker hub
 	docker tag local/kicbase:latest $(KIC_BASE_IMAGE_HUB)
 	$(MAKE) push-docker IMAGE=$(KIC_BASE_IMAGE_HUB)
 
-.PHONY: push-kic-base-image
-push-kic-base-image: ## Push local/kicbase:latest to all remote registries
+.PHONY: push-kic-base-image-x86
+push-kic-base-image-x86: ## Push local/kicbase:latest to all remote registries
```
Comment (medyagh) [inline Makefile:679]: is this still needed ? or is it used by anything ?

**Review by medyagh — APPROVED** [review]
thank you for the good work on this PR, feel free to merge after docker_linux tests are normal

**Pull request merged**